doc/dbus-interfaces.md

```diff
@@ -0,0 +1,328 @@
+# Hierte interfaces
+
+Hirte is a tool that allows you to manage a multitude of systems, each
+running systemd, in a single place.
+
+The main service in hierte is called the orchestrator. It runs on some
+kind of linux system, that linux system itself may be connected to
+hirte itself, but that is not strictly necessary.
+
+When the orchestrator starts it loads a configuration file that
+describes all the systems (called nodes) that are to be
+orchestrated. Each node has a unique name that is used to reference it
+in the orchestrator. Optionally the configuration also contains
+authentication information for each node (using client certificates).
+
+On each node that is under control of hirte a service called
+"hirte-node" is running, when it starts up it connects (via dbus over
+TCP) to the orchestrator and registers as available (optionally
+authenticating). It then receives reqests from the orchestrator and
+reports locate state changes to it.
+
+The main way to interact with Hirte is using DBus. It exposes a name
+on the system bus called `org.containers.hirte` that other programs
+can use to control the system. It is expected that highlevel control
+planes use this API directly, but there is also a hierte-client
+program that is useful for debugging and testing.
+
+# Hirte public DBus API
+
+The main entry point is at the /org/containers/hirte object path and
+implements the "org.containers.hirte.Manager" interface.
+
+Note that all properties also come with chenge events, so you can
+easily track when they change.
+
+## interface org.containers.hirte.Manager
+
+* Methods:
+  * ListUnits(out a(sssssssouso) units)
+
+    Returns an array with all currently loaded systemd units on all
+    (online) nodes. This is equivalent to calling Node.ListUnits() on
+    all the online nodes and adding the name of the node as the first
+    element of each returned element struct.
+
+  * CreateMonitor(out o monitor)
+
+    Creates a new monitor object, which can be used to monitor the
+    state of selected units on the nodes. The monitor object returned
+    will automatically be closed if the calling peer disconnects from
+    the bus.
+
+* Signals:
+  * JobNew(u id,
+           o job,
+           s nodeName,
+           s unit)
+
+    Emitted each time a new hirte job is queued.
+
+  * JobRemoved(u id,
+               o job,
+               s nodeName,
+               s unit,
+               s result)
+
+    Emitted each time a new job is dequeued or the underlying systemd
+    job finished. `result` is one of: "done", "failed", "cancelled",
+    "timeout", "dependency", "skipped". This is either the result
+    from systemd on the node, or "cancelled" if the job was cancelled
+    in hirte before any systemd job was started for it.
```

Comment:
what should be returned if the job is cancelled while Hirte is iterating through the nodes? (so some nodes started on the job, others didn't)

Reply:
This is the hirte level job, so it will only be removed when it is completely completed in hirte. Right now no operations actually cause operations on multiple nodes, as we don't have a StartOnAllNodes() kind of operation. However, if we add one, this is something we have to be careful about.